### PR TITLE
Nibabel recipe.

### PR DESCRIPTION
**Pull request by arokem**

### CODE REVIEW

File: recipes/nibabel/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    - python
+    - numpy
+    - scipy
+test:
+  requires:
+    - nose
+  imports:
+    - nibabel
+about:
+  home: https://nipy.org/dipy
```
Comment (jakirkham) [inline recipes/nibabel/meta.yaml]: Should this be ( http://nipy.org/nibabel/ )?

File: recipes/nibabel/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    git_url: https://github.com/nipy/nibabel.git
+    git_tag: 2.0.2
+
+requirements:
+  build:
+    - python
+    - numpy
+  run:
+    - python
+    - numpy
```
Comment (jakirkham) [inline recipes/nibabel/meta.yaml]: Is this linking against NumPy?

File: recipes/nibabel/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package:
+  name: nibabel
+  version: "0.11"
+
+source:
+    git_url: https://github.com/nipy/nibabel.git
+    git_tag: 2.0.2
```
Comment (jakirkham) [inline recipes/nibabel/meta.yaml]: Is there a tarball we could use instead?

File: recipes/nibabel/meta.yaml (hunk 1 above)
Comment (arokem) [inline recipes/nibabel/meta.yaml]: It should! Thanks for spotting that.

File: recipes/nibabel/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
+package:
+  name: nibabel
+  version: "0.11"
```
Comment (jakirkham) [inline recipes/nibabel/meta.yaml]: Seems not to match the tag we are using below.

File: recipes/nibabel/build.sh [hunk 5]
```diff
@@ -0,0 +1 @@
+$PYTHON setup.py install     # Python command to install the script.
```
Comment (jakirkham) [inline recipes/nibabel/build.sh]: As it appears we could be using `setuptools`, could you please add these arguments?

     --single-version-externally-managed --record record.txt

File: recipes/nibabel/bld.bat [hunk 6]
```diff
@@ -0,0 +1,2 @@
+"%PYTHON%" setup.py install
```
Comment (jakirkham) [inline recipes/nibabel/bld.bat]: Also add the arguments here too.

     --single-version-externally-managed --record record.txt

File: recipes/nibabel/meta.yaml (hunk 2 above)
Comment (jakirkham) [inline recipes/nibabel/meta.yaml]: Guessing not as it appears to be pure Python. Could you please confirm?

File: recipes/nibabel/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  requires:
+    - nose
+  imports:
+    - nibabel
+about:
+  home: https://nipy.org/nibabel
+  license: MIT
+  license_file: COPYING
+  summary: Python package to access a cacophony of neuro-imaging file formats
+extra:
```
Comment (jakirkham) [inline recipes/nibabel/meta.yaml]: Could we please space out these sections? One newline before `extra` should be fine.

File: recipes/nibabel/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  run:
+    - python
+    - numpy
+    - scipy
+test:
+  requires:
+    - nose
+  imports:
+    - nibabel
+about:
```
Comment (jakirkham) [inline recipes/nibabel/meta.yaml]: Same here.

File: recipes/nibabel/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+requirements:
+  build:
+    - python
+    - numpy
+  run:
+    - python
+    - numpy
+    - scipy
+test:
```
Comment (jakirkham) [inline recipes/nibabel/meta.yaml]: Also here.

File: recipes/nibabel/meta.yaml (hunk 2 above)
Comment (arokem) [inline recipes/nibabel/meta.yaml]: I believe it does require numpy: https://github.com/nipy/nibabel/blob/master/requirements.txt

Comment (jakirkham) [inline recipes/nibabel/meta.yaml]: Sure, what I meant was is there compiled C/C++ or other languages that build libraries that link to NumPy libraries. It doesn't appear that is the case from the code I have seen. Is that a correct statement?

Comment (arokem) [inline recipes/nibabel/meta.yaml]: And there are some modules that require scipy, e.g.: https://github.com/nipy/nibabel/blob/master/nibabel/processing.py

File: recipes/nibabel/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+  requires:
+    - nose
+  imports:
+    - nibabel
+about:
+  home: https://nipy.org/dipy
+  license: MIT
+  license_file: COPYING
+  summary: Python package to access a cacophony of neuro-imaging file formats
+extra:
```
Comment (jakirkham) [inline recipes/nibabel/meta.yaml]: Also, another newline here above `extra`.

File: recipes/nibabel/meta.yaml (hunk 2 above)
Comment (arokem) [inline recipes/nibabel/meta.yaml]: Sorry - didn't understand the question at first. That's correct.

File: recipes/nibabel/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,32 @@
+package:
+  name: nibabel
+  version: "2.0.2"
+
+source:
+  fn: nibabel-2.0.2.tar.gz
+  url: https://pypi.python.org/packages/source/n/nibabel/nibabel-2.0.2.tar.gz
+  md5: fa89163149e35f52e237b2cb56b1c606
```
Comment (jakirkham) [inline recipes/nibabel/meta.yaml]: Some people like to use jinja templates for things that get repeated often like version numbers. It makes it much easier to update. Here is an [example]( https://github.com/conda-forge/megaman-feedstock/blob/master/recipe/meta.yaml#L1-L9 ).

File: recipes/nibabel/meta.yaml (hunk 2 above)
Comment (jakirkham) [inline recipes/nibabel/meta.yaml]: Ok, thanks for clarifying. This is fine as is then.

File: recipes/nibabel/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+requirements:
```
Comment (jakirkham) [inline recipes/nibabel/meta.yaml]: Above `requirements` add a build section like [this]( https://github.com/conda-forge/django-extensions-feedstock/blob/master/recipe/meta.yaml#L10-L12 ). Also, delete `build.sh`, but keep `bld.bat`.

**Comment by jakirkham** [comment]
Weird. So, it seems it is using `distutils`. I was a little thrown by the refs I saw in the `setup.py` to `setuptools`. Feel free to strip the extra arguments from the `python setup.py install`s. That should clear up this build issue.

**Comment by arokem** [comment]
Thanks for all the help (and for making this!)

Do you have a preference: `build` section in the meta file vs. separate
`build.sh` file?

**Comment by jakirkham** [comment]
> Thanks for all the help...

Absolutely.

> ...(and for making this!)

Well, I can't really take credit for that. :blush:

> Do you have a preference: `build` section in the meta file vs. separate `build.sh` file?

We prefer the `build` section if it is one line. Otherwise `build.sh` is used. We would eliminate the `bld.bat` too, but `conda build` has a bug that makes that unfeasible.

File: recipes/nibabel/bld.bat (hunk 6 above)
Comment (jakirkham) [inline recipes/nibabel/bld.bat]: Not necessary as we discovered it uses `distutils`.

File: recipes/nibabel/build.sh (hunk 5 above)
Comment (jakirkham) [inline recipes/nibabel/build.sh]: Not necessary as we discovered it uses `distutils`.

**Comment by jakirkham** [comment]
As long as it passes CI, I am :+1: on merging. @ocefpaf, do you have anything you want to add?

**Comment by jakirkham** [comment]
Hmm...seems that this may require `h5py`. Is that a test requirement only or does the package uses this in its code too?

**Comment by jakirkham** [comment]
Looks like `h5py` is an optional dependency, but there is really no point in excluding it. Could you please add `h5py` to the build and run requirements? Should fix this test failure.

**Comment by ocefpaf** [comment]
LGTM :+1:

**Comment by arokem** [comment]
Is the right thing here to go upstream and create the `pydicom` recipe first?

**Comment by jakirkham** [comment]
Depends. There are two options.

Add a new PR with `pydicom`. Once that is merged and released, we can restart the builds here.

Alternatively, you can add it to this PR. Once merged both will release, but they may be out of sync with their builds. Depends on how you would like to do your work and how much you like pushing the CI restart buttons. :smile:

**Comment by arokem** [comment]
I'll go ahead and make a pydicom PR first. Will also be a good test for
whether I've learned from this :-)

**Comment by jakirkham** [comment]
Sounds good. Looking forward to it. :smile:

**Comment by arokem** [comment]
OK, once #286 gets merged, we can move ahead on this.

**Comment by ocefpaf** [comment]
@arokem any idea about what is going on with Windows and Python 3.5? Compiler/Features?

**Comment by jakirkham** [comment]
Yeah, it needs features. Please take a look at [this]( https://github.com/conda/conda/wiki/VC-features ), @arokem, and update this respectively.

**Comment by arokem** [comment]
Is there a good example here somewhere that I could follow?

**Comment by ocefpaf** [comment]
Sure all you need to do is to add the features field in the `build` section.

```
build:
    features:
        - vc9  # [win and py27]
        - vc10  # [win and py34]
        - vc14  # [win and py35]
```

**Comment by arokem** [comment]
I see this with a hash in some places (here) and without a hash in other
places (e.g,
https://github.com/conda/conda-recipes/pull/362#issuecomment-149197350).
Either way -- doesn't work. Anyone see what's wrong here?

File: recipes/nibabel/meta.yaml [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+build:
+    number: 0
+    script: python setup.py install
+
+requirements:
+  build:
+      features:
+          - vc9     [win and py27]
+          - vc10     [win and py34]
+          - vc14      [win and py35]
```
Comment (jakirkham) [inline recipes/nibabel/meta.yaml]: Move these from `requirements/build` to the `build` section above. Also, add `#` in front of each selector with a space between the two. Would be nice if they were indented to match.

**Comment by jakirkham** [comment]
Yep, added some notes in the diff that should help us fix this.

File: recipes/nibabel/run_test.py [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+import re
+
+splatform = sys.platform
+if splatform.startswith('win32'):
+    # Installation of data files for script testing a bit rickety on Appveyor:
+    config = nose.config.Config(verbosity=2,
+                                exclude=[re.compile("test_scripts"),
+                                         re.compile("test_dft")])
+else:
+    config = nose.config.Config(verbosity=2)
```
Comment (pelson) [inline recipes/nibabel/run_test.py]: Would you mind taking the verbosity down? The logs are a bit too long for the CI services to render, and all the conda recipe should really be doing is verifying the build was good, not acting as extra coverage for the package itself so we don't really need to know *which* tests are passing/failing until we want to debug it.

Comment (arokem) [inline recipes/nibabel/run_test.py]: Verbosity reduced! The nose output now seem to be going off the right side of the logs on Appveyor :-/

Failures remain only on Python 3.5/Windows. Any ideas?

Comment (lesteve) [inline recipes/nibabel/run_test.py]: > Failures remain only on Python 3.5/Windows. Any ideas?

It looks like there are genuine failures (You can E and F in the nosetests output) but for some reason it looks like the AppVeyor run was stopped before completing (there is no nosetests summary output showing the actual failures). Can you try doing a `git commit --amend` + force-push to trigger a new run of the CIs and double-check whether this was not a temporary glitch?

Comment (arokem) [inline recipes/nibabel/run_test.py]: Thanks for taking a look. Just force pushed. Fingers crossed...

Comment (lesteve) [inline recipes/nibabel/run_test.py]: Looks like genuine failures that only happen with python 3.5 for some reason and at least the AppVeyor log now has the full output of nosetests ...

Comment (arokem) [inline recipes/nibabel/run_test.py]: I just noticed that the nipy/nibabel repo is not being tested against python 3.5 on this platform...

For example: https://ci.appveyor.com/project/Eric89GXL/nibabel/build/1.0.603

I'll make a PR on nibabel to test on 3.5, so we can see whether we get the same failures there, or if it's something idiosyncratic happening here.

**Comment by arokem** [comment]
OK - after some investigation, I have come to the conclusion that these 3.5 failures are not a general failure of the original package (tests pass on Python 3.5 on Appveyor here: https://github.com/nipy/nibabel/pull/446), and are probably related to specifics of this setup. In particular, it looks like Appveyor is running out of disk space, required for some tests (this is a data IO library, after all...). Any ideas how to make progress here? Should I reduce testing on 3.5/Windows?

**Comment by pelson** [comment]
> Should I reduce testing on 3.5/Windows?

Sure. In many situations, just testing that the package imports is a fair compromise. On general, it isn't possible for conda-forge to build then run every unit test for every package, so we need to find the sweet spot which minimised cycles on the CI services, but which also gives us good confidence that the package has been built appropriately.

In truth, the only people with enough experience of the package to do this is the maintainers themselves, so it will have to be a judgement call from you as to what is appropriate level of testing here.

**Comment by jakirkham** [comment]
LGTM. Thought?

**Comment by lesteve** [comment]
I'd be +1 to merge this one and go without running the test suite for now.

The nibabel Python 3.5 problem on Windows seems genuine and not that trivial to fix. I was able to track down on of the failure to numpy see https://github.com/numpy/numpy/issues/7601 for more details.

**Comment by jakirkham** [comment]
I agree. This has sat long enough. Please direct further comments to the feedstock.

**Pull request merged**